Support HEIC for previews

steiny2k opened the pull request:
Adapt https://github.com/owncloud/core/pull/30093 to NextCloud to tackle https://github.com/nextcloud/server/issues/7406

Testcase is developed, please review.

steiny2k committed:
Develop HEIC/HEIF preview support #7406

Signed-off-by: Sebastian Steinmetz <me@sebastiansteinmetz.ch>

steiny2k committed:
Testcase for HEIC image (similar to the ImageTest)

Signed-off-by: Sebastian Steinmetz <me@sebastiansteinmetz.ch>

steiny2k commented:
Testcases are failing in CI environment, as ImageMagick probably doesn't handle HEIC format. If I run the test in my machine I see all green:

![image](https://user-images.githubusercontent.com/462714/43985377-4e17caaa-9d07-11e8-9a9a-8b5e1e0e4024.png)

If possible, please update ImageMagick (I'm running 7.0.8-9) and compile it with config option `--with-libheif`. Make sure to have libde265 available so that it can find the necessary library. In macOS it's quite convenient using homebrew:
`brew install --with-libheif imagemagick `

rullzer commented:
Let me see what I can do here

rullzer commented:
Ah so updating the containers with a new imagick is not trivial... but top of that let me make some comments.

rullzer commented on lib/private/Preview/HEIC.php:
```diff
@@ -0,0 +1,32 @@
+namespace OC\Preview;
+
+class HEIC extends Bitmap {
+	/**
```
Should also override the isAvailable method. That can then check for support in imagemagick. Else this will do :boom: on system that do now have support for it.

rullzer commented on tests/lib/Preview/HEICTest.php:
```diff
@@ -0,0 +1,42 @@
+class HEICTest extends Provider {
+
+	public function setUp() {
+		parent::setUp();
```
Since we also run CI on older systems. A check here would also be good. Then we can add imagick support in a newer version. And skip it here if the user running it has an unsupported versions.

rullzer commented:
@steiny2k sorry for having this around so long. Thnx for the PR :)

steiny2k committed:
Address comments from @rullzer:
 - implement isAvailable
 - run tests only if ImageMagick with HEIC support is available in the
   environment

Signed-off-by: Sebastian Steinmetz <me@sebastiansteinmetz.ch>

steiny2k commented:
Hi @rullzer, I think I've addressed your comments with the last input. Would you mind having another look?

rullzer commented:
@steiny2k sorry for having this around for a while.

yes I think JPG makes more sense indeed.

steiny2k commented:
Hi @rullzer thanks, I've been looking into jpg over png, and would submit it tonight. However it would need to change the inheritance from Bitmap to the more Provider class, and copy&pasting some portions of the logic in Bitmap over. Or would you rather change the Bitmap class? It would basically mean to change the private functions `getResizedPreview` and `resize` to be protected so I can overwrite them?

Thanks for your input.

Also found an issue with the isAvailable check.

MorrisJobke reviewed (approved):
👍

MorrisJobke commented:
Oh - just have seen the update now ... Could you submit this as a separate PR then? :)

MorrisJobke commented:
Thanks for this nice contribution 👍

MorrisJobke commented:
If you are looking for other nice first tasks you can have a look at https://github.com/nextcloud/server/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22 :)

We are happy to help if you have questions.
